_posts/2021-08-13-multi-factor-authentication-with-twilio-verify-api.md

```diff
@@ -0,0 +1,88 @@
+
+1. Intervene when the user attempts to login by requesting a valid code. This is sent via SMS or through an automated phone call.
+2. Use Twilio Verify API to check the validity of the extra factor being provided i.e. that the code entered by the user during the login process is the one provided by the API.
+3. If the Verify API does not deem the code entered to be valid, do not allow the user to proceed with the login process.
+4. Give the option to verify their login attempt again.
+5. Unless the user can verify their login attempt via the second step, do not allow access to the account.
+<br><br>
+
+### Example of a simple Twilio integration for sending and verifying SMS verification code
+
+Upon creating an account with Twilio, obtain the account ID as well as the authentication token from the Twilio Console. Navigate to Twilio Verify in the console, and create a verify service. There is a graphical user interface here that makes the creation of the verify service quick and easy. Upon completing this step, you will be provided with an identification code for the service.
+
+These credentials should be stored in the secrets file as they are sensitive information that may be exploited. Install the ‘twilio-ruby’ gem in order to ensure access to the Twilio specific methods.
+
+Create and send a verification through SMS or phone call to the phone number being verified. You can change the verification method in the channel parameter.
+
+```ruby
+require 'rubygems'
+require 'twilio-ruby'
+account_sid = ENV['TWILIO_ACCOUNT_SID']
+auth_token = ENV['TWILIO_AUTH_TOKEN']
+service_id = ENV['TWILIO_SERVICE']
+@client = Twilio::REST::Client.new(account_sid, auth_token)
+
+verification = @client.verify
+               .services(service_id)
+               .verifications
+               .create(to: params[:phone_number_to_verify], channel: 'sms')
+```
+<br><br>
+Then use the same Twilio client and service to create a verification check with the code that the user has entered.
+
+```ruby
+verification_check = @client.verify
+                     .services(service_id)
+                     .verification_checks
+                     .create(to: params[:phone_number_to_verify], code: params[:code_entered])
+```
+<br><br>
+The verification check created has a status which will either equal ‘approved’ when the correct code is entered, or ‘pending’ when the incorrect code is entered or if it has not been provided by the user yet.
+
+This very simple integration can be used to support the high-level steps that allow the use of SMS as an additional factor for the authentication process, thus allowing users access to greater levels of security for their accounts.
+
+All in all, it was a great experience to be able to focus on the complexities of handling the process of 2SA. The Twilio Verify API lightened our workload as it’s much simpler in comparison to other providers of automated SMS messages and phone calls for the set up of MFA.
+
```

Comment:
Prehaps some final comment the summeries the benifits to users. 

Multi step authentiation is a simple way to vastly increase the security of your online accounts. In the coming weeks, we will be releasing this for all OnePageCRM users. Watch this space! 

Or something to this effect